plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java

```diff
@@ -98,11 +90,20 @@ public InputStream readBlob(String blobName) throws IOException {
             throw new NoSuchFileException("Blob [" + blobName + "] does not exist");
         }
         // FSDataInputStream does buffering internally
-        return store.execute(new Operation<InputStream>() {
-            @Override
-            public InputStream run(FileContext fileContext) throws IOException {
-                return fileContext.open(new Path(path, blobName), bufferSize);
-            }
+        return store.execute((Operation<InputStream>) fileContext -> {
+            FSDataInputStream is = fileContext.open(new Path(path, blobName), bufferSize);
+            return new InputStream() {
+                @Override
+                public int read() throws IOException {
+                    try {
+                        SpecialPermission.check();
+                        // FSDataInputStream can open connection on read()
+                        return AccessController.doPrivileged((PrivilegedExceptionAction<Integer>) is::read);
```

Comment:
can we maybe try to trigger this differently? I mean can we for instance try to call `#available()` or can we maybe read the first byte on open and wrap in a `BufferedInputStream` and then do this:

```Java

InputStream stream = is.isMarkSupported() ? is : new BufferedInputStream(is);
// do the following in doPrivileged?
stream.mark(1);
stream.skip(1);
stream.reset();
return stream;
```